src/java.base/share/classes/java/util/zip/Deflater.java

```diff
@@ -872,14 +872,41 @@ public void reset() {
      * Closes the compressor and discards any unprocessed input.
      *
      * This method should be called when the compressor is no longer
-     * being used. Once this method is called, the behavior of the
-     * Deflater object is undefined.
+     * being used. Once this method is called, further operations using
+     * this Deflater may throw an exception.
+     *
+     * @see #close()
      */
     public void end() {
         synchronized (zsRef) {
+            // check if already closed
+            if (zsRef.address() == 0) {
+                return;
+            }
             zsRef.clean();
             input = ZipUtils.defaultBuf;
+            inputArray = null;
+        }
+    }
+
+    /**
+     * Releases resources held by this compressor and discards any unprocessed input.
+     * This method should be called when the compressor is no longer needed.
+     *
+     * @implNote This method calls the {@link #end()} method. This method is a no-op
+     * if this compressor has already been previously closed, either through {@code close()}
+     * or {@code end()}
+     * @since 24
+     */
+    @Override
+    public void close() {
+        synchronized (zsRef) {
+            // check if already closed
```

Comment:
Should we comment `// in case subclasses override the closed check in end()` instead? Otherwise the duplicate comments and checks are confusing.

Reply:
Hello Chen, we want the close() to be idempotent irrespective of whether or not end() is overridden. That check in close() and the code comment on that line is to indicate that. If the comment is adding to confusion, I can remove it - I don't think it's a must to have it to understand the check.

Reply:
I mean, we have updated the default implementation of end() to be idempotent. It now just appears confusing to readers that close() **seems to** perform redundant checks for idempotence until they realize subclasses can override end() to make it non-idempotent.

My suggestion is that our comment should say why we have this check here even though there's already an identical check in end().

Reply:
> My suggestion is that our comment should say why we have this check here even though there's already an identical check in end().

I now see what you mean. I'll update this appropriately tomorrow.